modules/brew/brew.sh

```diff
@@ -0,0 +1,234 @@
+ExecStart=/usr/bin/cp -R -n /tmp/homebrew/home/linuxbrew/.linuxbrew /var/home/linuxbrew
+ExecStart=/usr/bin/chown -R 1000:1000 /var/home/linuxbrew
+ExecStart=/usr/bin/rm -rf /tmp/homebrew
+ExecStart=/usr/bin/touch /etc/.linuxbrew
+
+[Install]
+WantedBy=default.target multi-user.target
+EOF
+
+cat >/usr/lib/systemd/system/brew-update.service <<EOF
+[Unit]
+Description=Auto update brew for mutable brew installs
+After=local-fs.target
+After=network-online.target
+ConditionPathIsSymbolicLink=/home/linuxbrew/.linuxbrew/bin/brew
+
+[Service]
+User=1000
+Type=oneshot
+Environment=HOMEBREW_CELLAR=/home/linuxbrew/.linuxbrew/Cellar
+Environment=HOMEBREW_PREFIX=/home/linuxbrew/.linuxbrew
+Environment=HOMEBREW_REPOSITORY=/home/linuxbrew/.linuxbrew/Homebrew
+ExecStart=/usr/bin/bash -c "/home/linuxbrew/.linuxbrew/bin/brew update"
+EOF
+
+cat >/usr/lib/systemd/system/brew-upgrade.service <<EOF
+[Unit]
+Description=Upgrade Brew packages
+After=local-fs.target
+After=network-online.target
+ConditionPathIsSymbolicLink=/home/linuxbrew/.linuxbrew/bin/brew
+
+[Service]
+User=1000
+Type=oneshot
+Environment=HOMEBREW_CELLAR=/home/linuxbrew/.linuxbrew/Cellar
+Environment=HOMEBREW_PREFIX=/home/linuxbrew/.linuxbrew
+Environment=HOMEBREW_REPOSITORY=/home/linuxbrew/.linuxbrew/Homebrew
+ExecStart=/usr/bin/bash -c "/home/linuxbrew/.linuxbrew/bin/brew upgrade"
+EOF
+
+# Write systemd timer files dynamically
+cat >/usr/lib/systemd/system/brew-update.timer <<EOF
+[Unit]
+Description=Timer for brew update for mutable brew
+Wants=network-online.target
+
+[Timer]
+OnBootSec=${WAIT_AFTER_BOOT_UPDATE}
+OnUnitInactiveSec=${UPDATE_INTERVAL}
+Persistent=true
+
+[Install]
+WantedBy=timers.target
+EOF
+
+cat >/usr/lib/systemd/system/brew-upgrade.timer <<EOF
+[Unit]
+Description=Timer for brew upgrade for on image brew
+Wants=network-online.target
+
+[Timer]
+OnBootSec=${WAIT_AFTER_BOOT_UPGRADE}
+OnUnitInactiveSec=${UPGRADE_INTERVAL}
+Persistent=true
+
+[Install]
+WantedBy=timers.target
+EOF
+
+# Copy shell configuration files
+cp -r "${MODULE_DIRECTORY}"/brew/brew-fish-completions.fish /usr/share/fish/vendor_conf.d/brew-fish-completions.fish
+cp -r "${MODULE_DIRECTORY}"/brew/brew-bash-completions.sh /etc/profile.d/brew-bash-completions.sh
+
+# Register path symlink
+# We do this via tmpfiles.d so that it is created by the live system.
+cat >/usr/lib/tmpfiles.d/homebrew.conf <<EOF
+d /var/lib/homebrew 0755 1000 1000 - -
+d /var/cache/homebrew 0755 1000 1000 - -
+d /var/home/linuxbrew 0755 1000 1000 - -
+EOF
+
+# Enable the setup service
+systemctl enable brew-setup.service
+
+# Always enable or disable update and upgrade services for consistency
+if [[ "${AUTO_UPDATE}" == true ]]; then
+    systemctl enable brew-update.timer
+else
+    systemctl disable brew-update.timer
+fi
+
+if [[ "${AUTO_UPGRADE}" == true ]]; then
+    systemctl enable brew-upgrade.timer
+else
+    systemctl disable brew-upgrade.timer
+fi
+
+# Apply nofile limits if enabled
+if [[ "${NOFILE_LIMITS}" == true ]]; then
```

Comment:
I have an idea on how to improve this process.

We can check the current nofile limit in the base image with:

SSH/tty (not sure if it shows SystemD limit even when not set in limits.d, needs to be tested):
`ulimit -Sn` (soft limit)
`ulimit -Hn` (hard limit)

SystemD:
`systemctl show -all`
(`DefaultLimitNOFILE` = hard limit, `DefaultLimitNOFILESoft` = soft limit)

If the base image has limit less than desired, than do the change.
If it's equal or has more than enough, than do nothing & inform user that no change is needed.

Reply:
This is what I came up with.

It's a 231 lines of code, so `source` would be used below `nofile-limits` if statement.

It dynamically searches for CURRENT soft & hard limit between these files & directories, sorted by preference (from least preferred to most preferred):

SSH/TTY limits:
```
"/usr/etc/security/limits.conf"
"/usr/etc/security/limits.d/"
"/etc/security/limits.conf"
"/etc/security/limits.d/"
```

SystemD limits:
```
# System
"/usr/lib/systemd/system.conf"
"/usr/lib/systemd/system.conf.d/"
"/usr/etc/systemd/system.conf"
"/usr/etc/systemd/system.conf.d/"
"/etc/systemd/system.conf"
"/etc/systemd/system.conf.d/"

# User
"/usr/lib/systemd/user.conf"
"/usr/lib/systemd/user.conf.d/"
"/usr/etc/systemd/user.conf"
"/usr/etc/systemd/user.conf.d/"
"/etc/systemd/user.conf"
"/etc/systemd/user.conf.d/"
```

It always shows the CURRENT most preferred soft & hard nofile value.

OPTIMAL nofile limit values are set on these conditions:
- If CURRENT value is equal or more than OPTIMAL, than nothing is done.
- If CURRENT value is less than OPTIMAL, changes are made
- It accounts for the scenario when soft value is OPTIMAL, while hard value is not & vice-versa (so only 1st value is necessarily overwritten, while 2nd value remains unchanged)

What is not took into account:
- `limits.d` limit applies nofile 

File is put in `/usr/lib/systemd/{system,user}.conf.d/`, while `/usr/etc/security/limits.d/` must be used, since `/usr/lib` directory for this doesn't exist.

File is named `zz1-brew-limits.conf`. The reason why it has `zz1-` prefix is to assure that the file will be in highest priority. This takes advantage of lexic sorting, which is preferred in terms of priorities.
While systemd & others recommend using numbers as prefix, we can't guarantee that some image-maintainer doesn't have `99-limits.conf` file, which would override our file.
But most image-maintainers don't know about lexic sorting & using letters as prefixes being possible, so I guess that no image-maintainer at the moment has `zz1-` prefix or higher.

The potential issue can come where image-maintainer applies the modification to `/usr/etc` for systemd modifications. If he modified nofile limits, than his configuration will have higher priority.
However, not a lot of users do this I believe, so just a nitpick.

What do y'all think?

Here's the script:

```bash
#!/usr/bin/env bash

set -euo pipefail

# Script used to read nofile limits from the base image, since issuing easy systemctl status commands don't work in build-time.
# Takes into account config priorities & current config settings regarding nofile limits
# (if nofile limit is already applied with appropriate values in the base image, than this tweak is not applied)
# Modifies limits nofile value & systemd DefaultLimitNOFILE value

DESIRED_SOFT_LIMIT=4096
DESIRED_HARD_LIMIT=524288

# SSH/TTY nofile limit (security ulimit config)

# From least to most preferred
SSH_TTY_LIMIT_ORDER=(
"/usr/etc/security/limits.conf"
"/usr/etc/security/limits.d/"
"/etc/security/limits.conf"
"/etc/security/limits.d/"
)
updated_ssh_array=()
for path in "${SSH_TTY_LIMIT_ORDER[@]}"; do
    if [ -e "${path}" ]; then
        updated_ssh_array+=("${path}")
    fi
done
# Update the original array with the existing paths
SSH_TTY_LIMIT_ORDER=("${updated_ssh_array[@]}")

# Soft SSH/TTY nofile limit
SSH_TTY_SOFT_INFO=$(find "${SSH_TTY_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -v OFS='\t' '/soft\s+nofile/ && !/^#/ {sub(/.*nofile/, ""); gsub(/^[ \t]+/, ""); print FILENAME, $0}' {} + | tail -n 1)
if [[ -n "${SSH_TTY_SOFT_INFO}" ]]; then
  CURRENT_SSH_TTY_SOFT_VALUE=$(echo "${SSH_TTY_SOFT_INFO}" | awk '{print $2}')
else
  CURRENT_SSH_TTY_SOFT_VALUE=0
fi

# Hard SSH/TTY nofile limit
SSH_TTY_HARD_INFO=$(find "${SSH_TTY_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -v OFS='\t' '/hard\s+nofile/ && !/^#/ {sub(/.*nofile/, ""); gsub(/^[ \t]+/, ""); print FILENAME, $0}' {} + | tail -n 1)
if [[ -n "${SSH_TTY_HARD_INFO}" ]]; then
  CURRENT_SSH_TTY_HARD_VALUE=$(echo "${SSH_TTY_HARD_INFO}" | awk '{print $2}')
else
  CURRENT_SSH_TTY_HARD_VALUE=0
fi

# SystemD nofile limit

# SystemD system soft & hard nofile limit
# From least to most preferred
SYSTEMD_SYSTEM_LIMIT_ORDER=(
"/usr/lib/systemd/system.conf"
"/usr/lib/systemd/system.conf.d/"
"/usr/etc/systemd/system.conf"
"/usr/etc/systemd/system.conf.d/"
"/etc/systemd/system.conf"
"/etc/systemd/system.conf.d/"
)
updated_systemd_system_array=()
for path in "${SYSTEMD_SYSTEM_LIMIT_ORDER[@]}"; do
    if [ -e "${path}" ]; then
        updated_systemd_system_array+=("${path}")
    fi
done
# Update the original array with the existing paths
SYSTEMD_SYSTEM_LIMIT_ORDER=("${updated_systemd_system_array[@]}")

# SystemD system soft & hard nofile limit
SYSTEMD_SYSTEM_SOFT_INFO=$(find "${SYSTEMD_SYSTEM_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $2}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_SYSTEM_SOFT_INFO}" ]]; then
  CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE=$(echo "${SYSTEMD_SYSTEM_SOFT_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE=0
fi

SYSTEMD_SYSTEM_HARD_INFO=$(find "${SYSTEMD_SYSTEM_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $3}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_SYSTEM_HARD_INFO}" ]]; then
  CURRENT_SYSTEMD_SYSTEM_HARD_VALUE=$(echo "${SYSTEMD_SYSTEM_HARD_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_SYSTEM_HARD_VALUE=0
fi

# SystemD user soft & hard nofile limit
SYSTEMD_USER_LIMIT_ORDER=(
"/usr/lib/systemd/user.conf"
"/usr/lib/systemd/user.conf.d/"
"/usr/etc/systemd/user.conf"
"/usr/etc/systemd/user.conf.d/"
"/etc/systemd/user.conf"
"/etc/systemd/user.conf.d/"
)
updated_systemd_user_array=()
for path in "${SYSTEMD_USER_LIMIT_ORDER[@]}"; do
    if [ -e "${path}" ]; then
        updated_systemd_user_array+=("${path}")
    fi
done
# Update the original array with the existing paths
SYSTEMD_USER_LIMIT_ORDER=("${updated_systemd_user_array[@]}")

SYSTEMD_USER_SOFT_INFO=$(find "${SYSTEMD_USER_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $2}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_USER_SOFT_INFO}" ]]; then
  CURRENT_SYSTEMD_USER_SOFT_VALUE=$(echo "${SYSTEMD_USER_SOFT_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_USER_SOFT_VALUE=0
fi

SYSTEMD_USER_HARD_INFO=$(find "${SYSTEMD_USER_LIMIT_ORDER[@]}" -type f -name "*.conf" -exec awk -F'[:=]' '/^[^#]*DefaultLimitNOFILE/ {print FILENAME, $3}' {} + | tail -n 1)
if [[ -n "${SYSTEMD_USER_HARD_INFO}" ]]; then
  CURRENT_SYSTEMD_USER_HARD_VALUE=$(echo "${SYSTEMD_USER_HARD_INFO}" | awk '{print $2}')
else
  CURRENT_SYSTEMD_USER_HARD_VALUE=0
fi

# Check current state

echo "Current nofile limit values:"
check_and_print() {
    if [[ "${1}" -eq 0 ]]; then
        echo "UNSET"
    else
        echo "${1}"
    fi
}
echo "SSH/TTY soft nofile limit: $(check_and_print ${CURRENT_SSH_TTY_SOFT_VALUE})"
echo "SSH/TTY hard nofile limit: $(check_and_print ${CURRENT_SSH_TTY_HARD_VALUE})"
echo "SystemD system soft nofile limit: $(check_and_print ${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE})"
echo "SystemD system hard nofile limit: $(check_and_print ${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE})"
echo "SystemD user soft nofile limit: $(check_and_print ${CURRENT_SYSTEMD_USER_SOFT_VALUE})"
echo "SystemD user hard nofile limit: $(check_and_print ${CURRENT_SYSTEMD_USER_HARD_VALUE})"

# Write nofile limit values
# zz1- prefix is used for config, to assure that nofile limit is going to be applied, as it's high in lexical order.
# If config is higher in lexical order, that means that it's preferred over other ones.
# Downstreams can just go on with their own modifications in zz2-, zz3-, zz4-, zz5-, etc. increments
# if they wish to experiment with nofile limits differently

# Write SSH/TTY nolimit values
if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] || [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  if [[ ! -d "/usr/etc/security/limits.d/" ]]; then
    mkdir -p "/usr/etc/security/limits.d/"
  fi
  echo "# This file sets the resource limits for users logged in via PAM,
# more specifically, users logged in via SSH or tty (console).
# Limits related to terminals in Wayland/Xorg sessions depend on a
# change to /etc/systemd/user.conf.
# This does not affect resource limits of the system services.
# This file overrides defaults set in /etc/security/limits.conf

" > "/usr/etc/security/limits.d/zz1-brew-limits.conf"
fi

if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SSH/TTY soft nofile limit value"
  echo "* soft nofile ${DESIRED_SOFT_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]]; then
  echo "Required SSH/TTY soft nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SSH/TTY hard nofile limit value"
  echo "* hard nofile ${DESIRED_HARD_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Required SSH/TTY hard nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SSH_TTY_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SSH_TTY_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SSH/TTY soft & hard nofile limit value"
  echo "* soft nofile ${DESIRED_SOFT_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
  echo "* hard nofile ${DESIRED_HARD_LIMIT}" >> "/usr/etc/security/limits.d/zz1-brew-limits.conf"
fi

# Write SystemD nolimit values

# Writing SystemD system nolimit values
if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] || [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  if [[ ! -d "/usr/lib/systemd/system.conf.d/" ]]; then
    mkdir -p "/usr/lib/systemd/system.conf.d/"
  fi
fi

if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD system soft nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" > "/usr/lib/systemd/system.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]]; then
  echo "Required SystemD system soft nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD system hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/system.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Required SystemD system hard nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_SYSTEM_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_SYSTEM_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD system soft & hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/system.conf.d/zz1-brew-limits.conf"
fi

# Writing SystemD user nolimit values
if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] || [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  if [[ ! -d "/usr/lib/systemd/user.conf.d/" ]]; then
    mkdir -p "/usr/lib/systemd/user.conf.d/"
  fi
fi

if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD user soft nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${CURRENT_SYSTEMD_USER_HARD_VALUE}" > "/usr/lib/systemd/user.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]]; then
  echo "Required SystemD user soft nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -ge "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD user hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${CURRENT_SYSTEMD_USER_SOFT_VALUE}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/user.conf.d/zz1-brew-limits.conf"
elif [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -ge "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Required SystemD user hard nofile limit value is already satisfied!"
fi

if [[ "${CURRENT_SYSTEMD_USER_SOFT_VALUE}" -lt "${DESIRED_SOFT_LIMIT}" ]] && [[ "${CURRENT_SYSTEMD_USER_HARD_VALUE}" -lt "${DESIRED_HARD_LIMIT}" ]]; then
  echo "Writing increased SystemD user soft & hard nofile limit value"
  echo "[Manager]
DefaultLimitNOFILE=${DESIRED_SOFT_LIMIT}:${DESIRED_HARD_LIMIT}" > "/usr/lib/systemd/user.conf.d/zz1-brew-limits.conf"
fi
```

Reply:
But, this script is complex as it is, & with potential systemd changes, it could be broken.

So I would maybe lean more towards documenting this, even though I spent some time on this code.

Reply:
I merged it in.

I think that this is the best approach for this. It should be future-proof & reliable enough.